### ISSUE TITLE
Windows CRYPTO_THREAD_run_once() implementation does not guarantee memory ordering which leads to undefined behaviour in multithreading contexts

### ISSUE THREAD

**Issue by ioancea** [openssl/openssl#22059]
# Platform and versions
 * Windows 11 arm64
 * OpenSSL: 3.1.2 (but the current `master` is also affected)
 * MSVC 2022 compiler

# Description
The current implementation of `CRYPTO_THREAD_run_once()` for the Windows system (see https://github.com/openssl/openssl/blob/d2873946dfaff5537ea3d1adf3890e33a3f276ff/crypto/threads_win.c#L134-L152) has a subtle issue, which can be exploited especially when running on architectures with weak memory ordering CPUs (like ARM).

Function's lock-free loop correctly synchronises the access to the `once` variable (by doing atomic operation with `InterlockedCompareExchange` API), however if the `init()` function writes some kind of global/shared variable(s), it might lead to `out-of-order reads` by another threads - aka the "consumer" threads might see the global data as incomplete even if `once` was changed.
There are multiple reasons to lead into out-of-order memory access (e.g. compiler optimisations, CPU memory access reordering, CPU cache), see a good introduction to the topic https://learn.microsoft.com/en-us/windows/win32/sync/synchronization-and-multiprocessor-issues

In my case, I discovered the problem by investigating some sporadic crashes in `OPENSSL_thread_stop()` called in parallel from multiple threads, from an application which links OpenSSL statically. From time to time, it happened that `glob_tevent_reg` to be incomplete in `init_thread_remove_handlers` https://github.com/openssl/openssl/blob/d2873946dfaff5537ea3d1adf3890e33a3f276ff/crypto/initthread.c#L176 - aka `gtr->lock` was NULL which made https://github.com/openssl/openssl/blob/d2873946dfaff5537ea3d1adf3890e33a3f276ff/crypto/initthread.c#L179 dereference a null pointer.

Basically, with the current implementation of `CRYPTO_THREAD_run_once()` there's a real danger that MT applications to be affected w/o knowing where to pinpoint - e.g. the out-of-order data can easily create (heap) memory corruptions which are much more fun to debug than a crash by null pointer access.

# PoC and solutions
I have created a small project to show-case the problem and implemented some fix proposals - refer to the `README.md` file from the archive
[CRYPTO_THREAD_run_once-windows.zip](https://github.com/openssl/openssl/files/12574275/CRYPTO_THREAD_run_once-windows.zip)

My recommendation is to adapt `CRYPTO_THREAD_run_once ` implementation to call Windows's `InitOnceExecuteOnce` API when the target is post Windows Vista (especially because XP is EOL for a while) and to keep the current implementation (plus the `MemoryBarrier()` fix) for the obsolete XP builds.

**Comment by tom-cosgrove-arm** [comment]
Couldn't the implementation on Windows just use the `InitOnceExecuteOnce` Windows API designed to do "run once"?

**Comment by ioancea** [comment]
@tom-cosgrove-arm `InitOnceExecuteOnce` isn't available in XP which, according with the comment from the source code https://github.com/openssl/openssl/blob/d2873946dfaff5537ea3d1adf3890e33a3f276ff/crypto/threads_win.c#L130-L133, it is still supported by OpenSSL

**Comment by tom-cosgrove-arm** [comment]
It feels like a safer implementation with the correct native function would be better, with a compile-time fall-back if we don't have it. I'm not aware of WinXP running on systems with weak memory ordering

**Comment by ioancea** [comment]
Also, to be noticed that I mistakenly though MemoryBarrier (https://learn.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-memorybarrier#requirements) is available in Windows XP but that's actually Windows Vista+ too.

Thus @tom-cosgrove-arm 's suggestion to stick with the current implementation as fallback, looks legit to me

**Comment by nhorman** [comment]
but in the fallback case, the current implementation still suffers from the issue that, when multiple threads call CRYPTO_THREAD_run_once, threads above and beyond the first will still potentially return prior to the init routine completing, allowing for a write after read error of any global data modified by the init routine. It seems like a mutex is legitimately required here (which is what pthread_once does under the covers)

**Comment by hlandau** [comment]
> It feels like a safer implementation with the correct native function would be better, with a compile-time fall-back if we don't have it. I'm not aware of WinXP running on systems with weak memory ordering

Consider the strategy we take for ossl_crypto_condvar - we have cases there for NT-legacy and NT-modern. That's set at build time.

I suggest we do something similar here and leave this as a fallback for builds where _WIN32_WINNT is set to XP.

We could then change the default for _WIN32_WINNT to something more recent at least for ARM, since older builds of Windows for ARM literally don't exist.

Having said all of that, we should also fix this legacy code path. Can't we just add a memory barrier?

**Comment by ioancea** [comment]
> but in the fallback case, the current implementation still suffers from the issue that, when multiple threads call CRYPTO_THREAD_run_once, threads above and beyond the first will still potentially return prior to the init routine completing, allowing for a write after read error of any global data modified by the init routine. It seems like a mutex is legitimately required here (which is what pthread_once does under the covers)

Normally that shouldn't happen because the other threads will spin inside the loop until `ONCE_DONE` is set, which is done after the call to `init()`.
However, because there's no `happens-before relationship` between them, I think it's still possible (at least theoretically with aggressive optimisation) for the `compiler` to change the order and set `lock` before calling `init()`. So, indeed, a fix for the legacy path seems to be needed.

**Comment by mattcaswell** [comment]
Adding "help wanted" label

**Comment by ssbssa** [comment]
Instead of doing `*lock = ONCE_DONE;` directly, wouldn't [`InterlockedExchange`](https://learn.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedexchange) fix this (since it also generates a generates a full memory barrier)?

**Comment by t8m** [comment]
We should fix this. The plan is clear. And we should also bump the default supported WINNT version to Windows 7.

**Comment by nhorman** [comment]
Is there a reason we aren't using InitOnceExecuteOnce for this implementation?

**Comment by t8m** [comment]
@nhorman It is mentioned above among the comments - the minimum version of Windows we "support" is currently Win  XP.  We should drop that support and move on.

**Comment by nhorman** [comment]
@t8m so in attempting to make the conversion to InitOnceExecuteOnce here:
https://github.com/openssl/openssl/pull/31339

I found that it crashes because the Windows API requires that the CRYPTO_ONCE variable (aliased to an INIT_ONCE type) needs to be DWORD aligned, which is not clearly documented here:
https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-initonceexecuteonce
 
But seems to be common knoweldge on a a google search regarding usage.  There is no guarantee that global variables are going to be correctly aligned unless each global CRYPTO_ONCE variable is attributed with __declspec(align(sizeof(DWORD)))

That seems really messy.  Do you think its worth doing, or should we just clean up the implementation we have currently?

**Comment by tom-cosgrove-arm** [comment]
> That seems really messy. Do you think its worth doing, or should we just clean up the implementation we have currently?

If that alignment is known and a real requirement, then we should put it on the typedef

```
#if defined(_MSC_VER)
typedef __declspec(align(sizeof(DWORD))) LONG CRYPTO_ONCE;
#elif defined(__clang__) || defined(__GNUC__)
typedef LONG CRYPTO_ONCE __attribute__((aligned(sizeof(DWORD))));
#else
typedef LONG CRYPTO_ONCE;
#endif
```

**Comment by nhorman** [comment]
merged the fix to master.  Note that because this change relies on an api that is unavailable in Windows XP, it also required bumping our WINNT support version up, which we don't want to do in older branches.  As such, if you want to see this fixed in current stable releases, an alternate solution that does not rely on InitOnceExecuteOnce will need to be developed and new PR opened.

**Comment by t8m** [comment]
Unfortunately the fix breaks API/ABI compatibility. We will need to do something else.

**Comment by nhorman** [comment]
If we can't use InitOnceExecuteOnce, I think the only solution here is the memory barrier approach.  @ioancea can you please post  a patch with your proposed solution above using that method?

**Comment by ioancea** [comment]
@t8m @nhorman I did not understand the API/ABI compatibility issue with `InitOnceExecuteOnce` - bumping the internal `_WIN32_WINNT` does create the incompatibility? If so, this means OpenSSL will be forever stuck with the current, already obsolete, version? Or it's about `CRYPTO_THREAD_run_once` which now will expect another `once` type (from `LONG` to `INIT_ONCE`) ?

In my project, I simply failed OpenSSL compilation if `_WIN32_WINNT` is not at least `0x600` (Vista), and patched `threads_win.c` similar to https://github.com/openssl/openssl/commit/d053fe41a4add4e55dc5633e8aef6602a6a4b90b. For example, my `crypto.h.in` patch against the `3.5.6` version is:
```
--- include/openssl/crypto.h.in
+++ include/openssl/crypto.h.in
@@ -502,13 +502,21 @@

 #if defined(OPENSSL_THREADS) && !defined(CRYPTO_TDEBUG)
 #if defined(_WIN32)
+/* workaround to automatically include windows crypto defs _only_ when compiling OpenSSL */
+#   if defined(OPENSSL_BUILDING_OPENSSL)
+#    include <windows.h>
+#   endif
 #if defined(BASETYPES) || defined(_WINDEF_H)
 /* application has to include <windows.h> in order to use this */
 typedef DWORD CRYPTO_THREAD_LOCAL;
 typedef DWORD CRYPTO_THREAD_ID;

-typedef LONG CRYPTO_ONCE;
-#define CRYPTO_ONCE_STATIC_INIT 0
+#    if !defined(_WIN32_WINNT) || _WIN32_WINNT < 0x600
+#     error "Building with pre-Vista is not supported"
+#    endif
+typedef INIT_ONCE CRYPTO_ONCE;
+#    define CRYPTO_ONCE_STATIC_INIT INIT_ONCE_STATIC_INIT
+
 #endif
 #else
 #if defined(__TANDEM) && defined(_SPT_MODEL_)
```
 
If the support for XP is wanted, the above can be changed to set the old defines instead of throwing an error. And, of course, the `threads_win.c` implementation should be changed to support both.

**Comment by nhorman** [comment]
@ioancea the _WINNT_VER value isnt' the ABI breakage.  The breakage occurs because The arch specific run once variable for each architecture gets typedefed to the CRYPTO_ONCE type, which is statically declared by applications/libraries that use it.  CRYPTO_ONCE on windows is currently defined as LONG, and using InitOnceExecuteOnce requires changing the definition of CRYPTO_ONCE to an INIT_ONCE structure as defined in windows.h.  Thats a different size, and so applications built with openssl version X won't work properly with newer openssl versions that implement CRYPTO_THREAD_run_once with the InitOnceExecuteOnce API, because those static allocations will be of the wrong size.

In any case, we've fixed this with a memory barrier now.

**Comment by ioancea** [comment]
(Sorry for the double post, I removed the extra comment)
Many thanks for the clarification @nhorman. Maybe OpenSSL can break the API/ABI constraints with a major version then (in order to fix all the drama with this run_once and use `InitOnceExecuteOnce`)?

Even if I initially suggested the `MemoryBarrier` fix (done in #31713), I think it's incomplete (I could try to make another PoC for this part, if it's really necesary) because is not properly handling the "reader" side on the fast-exit branch (https://github.com/openssl/openssl/blob/cf82f5f78f57f56de49a5f346aea6a4baf020c7c/crypto/threads_win.c#L523). So, even if the MemoryBarrier ensures the writer publishes all the shared data before marking the lock `ONCE_DONE`, the reader threads can still load both of them out-of-order: it can see `once=ONCE_DONE` while the shared data is old (probably NULL, but not "incomplete" - because the writer memory barrier fixed that). To have a proper happens-before relationship, we need a release on the writer and acquire on the reader - this basically means the fast-exit check must be replaced with either a "real" acquire (like if (`InterlockedCompareExchange(lock, ONCE_DONE, ONCE_DONE) == ONCE_DONE)`) or dropped completely, as the rest of the loop should handle that part anyway.

Compared with my initial report, when the impact of the problem could have visible effects (like crashes, due to incomplete published data), the current fix will hide out-of-order reads because the "reader" usually has a `ptr = get_shared_data(); if (!ptr) -> abort_some_processing` (ptr can only be NULL if it's holding a stale value, in the moment `once=ONCE_DONE` on the reader). That in itself looks dangerous, even if the code has fail-safes in place, it's still a logic error which should be properly fixed.

**Comment by nhorman** [comment]
> (Sorry for the double post, I removed the extra comment) Many thanks for the clarification [@nhorman](https://github.com/nhorman). Maybe OpenSSL can break the API/ABI constraints with a major version then (in order to fix all the drama with this run_once and use `InitOnceExecuteOnce`)?
> 

Yes, that is exactly when we would have to make the switch

> Even if I initially suggested the `MemoryBarrier` fix (done in [#31713](https://github.com/openssl/openssl/pull/31713)), I think it's incomplete (I could try to make another PoC for this part, if it's really necesary) because is not properly handling the "reader" side on the fast-exit branch (
> 
> [openssl/crypto/threads_win.c](https://github.com/openssl/openssl/blob/cf82f5f78f57f56de49a5f346aea6a4baf020c7c/crypto/threads_win.c#L523)
> 
> Line 523 in [cf82f5f](/openssl/openssl/commit/cf82f5f78f57f56de49a5f346aea6a4baf020c7c)
> 
>  if (*lock == ONCE_DONE) 
> ). So, even if the MemoryBarrier ensures the writer publishes all the shared data before marking the lock `ONCE_DONE`, the reader threads can still load both of them out-of-order: it can see `once=ONCE_DONE` while the shared data is old (probably NULL, but not "incomplete" - because the writer memory barrier fixed that). To have a proper happens-before relationship, we need a release on the writer and acquire on the reader - this basically means the fast-exit check must be replaced with either a "real" acquire (like if (`InterlockedCompareExchange(lock, ONCE_DONE, ONCE_DONE) == ONCE_DONE)`) or dropped completely, as the rest of the loop should handle that part anyway.

I think we're actually ok here.  On the reader side, if a reader reads lock as ONCE_DONE, its implied that any data written via the init callback has been published, and (unless the windows MemoryBarrier() call does something really odd), the compiler shouldn't be allowed to re-order the read and write of the lock read and write operations accross the MemoryBarrier call.  In the worst case, a reader will read lock as ONCE_ININIT while the writer side has a pending write to lock (updating to ONCE_DONE), in which case that reader will fall into the do/while loop where it will get the synchronized read that you are looking for via InterlockedCompareExchange.

**Comment by ioancea** [comment]
@nhorman I think there's a misunderstanding, the issue isn't about the writer reordering (the `MemoryBarrier` fix handles that correctly). The issue is about the reader reordering its own loads - the `LoadLoad` reordering on ARM64. 

Consider two threads on ARM64:
```
Writer (Thread A):                         Reader (Thread B):
------------------                         ------------------ 
init() -> stores shared_data              load *lock -> sees ONCE_DONE
MemoryBarrier()                            return 1
*lock = ONCE_DONE                          load shared_data -> sees NULL
```

The MemoryBarrier in Thread A ensures Thread A's stores happen in order (`shared_data` before `DONE`). That's the release side, which is correct.

But Thread B does two loads - first `*lock`, then `shared_data`, with nothing between them. On ARM64, the CPU is allowed to satisfy those loads out of program order (LoadLoad reordering). Thread B can observe `DONE` from the updated cache line while `shared_data` still reads from a stale cache line. To make things worse, the CPU can also do `speculative/out-of-order load execution` which, unlike x86, there's no "snoop reply" - once the speculative load is satisfied from cache, it commits with that value even if an invalidation arrives afterward.

This cannot happen on x86/x64 (TSO forbids LoadLoad reordering), and it cannot happen if both values are on the `same cache line` (coherence delivers them atomically). It only manifests on weakly-ordered hardware (ARM64, POWER) when the flag and the data are on `different cache lines`, which they are (with high probability) in OpenSSL's case.

The fix then should be an acquire on the reader's fast-path, probably the cheaper being `MemoryBarrier` call - it's a full fence but is shouldn't have scalability issues (no exclusive cache line ownership):
```
if (*lock == ONCE_DONE) {
    MemoryBarrier();
    return 1;
}
```

**Comment by nhorman** [comment]
> [@nhorman](https://github.com/nhorman) I think there's a misunderstanding, the issue isn't about the writer reordering (the `MemoryBarrier` fix handles that correctly). The issue is about the reader reordering its own loads - the `LoadLoad` reordering on ARM64.
> 
> Consider two threads on ARM64:
> 
> ```
> Writer (Thread A):                         Reader (Thread B):
> ------------------                         ------------------ 
> init() -> stores shared_data              load *lock -> sees ONCE_DONE
> MemoryBarrier()                            return 1
> *lock = ONCE_DONE                          load shared_data -> sees NULL
> ```
> 
> The MemoryBarrier in Thread A ensures Thread A's stores happen in order (`shared_data` before `DONE`). That's the release side, which is correct.
> 
> But Thread B does two loads - first `*lock`, then `shared_data`, with nothing between them. On ARM64, the CPU is allowed to satisfy those loads out of program order (LoadLoad reordering). Thread B can observe `DONE` from the updated cache line while `shared_data` still reads from a stale cache line. To make things worse, the CPU can also do `speculative/out-of-order load execution` which, unlike x86, there's no "snoop reply" - once the speculative load is satisfied from cache, it commits with that value even if an invalidation arrives afterward.
> 
> This cannot happen on x86/x64 (TSO forbids LoadLoad reordering), and it cannot happen if both values are on the `same cache line` (coherence delivers them atomically). It only manifests on weakly-ordered hardware (ARM64, POWER) when the flag and the data are on `different cache lines`, which they are (with high probability) in OpenSSL's case.
> 
> The fix then should be an acquire on the reader's fast-path, probably the cheaper being `MemoryBarrier` call - it's a full fence but is shouldn't have scalability issues (no exclusive cache line ownership):
> 
> ```
> if (*lock == ONCE_DONE) {
>     MemoryBarrier();
>     return 1;
> }
> ```

I don't believe this is correct:
https://learn.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-memorybarrier

```
Creates a hardware memory barrier (fence) that prevents the CPU from re-ordering read and write operations. It may also prevent the compiler from re-ordering read and write operations.

```
That is to say, Any read/write that issues before the memory barrier must complete prior to any read/write that issues after the memory barrier.  Additional (though the ms learn page is soft on this), the compiler can't re-order instructions such that program order is modified across the barrier.

So you scenario:
|Writer (Thread A):        |                 Reader (Thread B): |
|---------------------------------|-----------------------------------------|
|init() -> stores shared_data| load *lock - > sees ONCE_DONE|
|MemoryBarrier()          |  return 1|
| *lock = ONCE_DONE | load share_data -> sees NULL|

cannot occur, because thread A, writing ONCE_DONE to *lock, may only complete after any prior issued stores/loads (as generated by the init pointer call) before the memory barrier have completed.  Thread B can only read a value of ONCE_DONE from lock until some arbitrary point after thread A has written it.  Therefore, transitively, thread B will only read ONCE_DONE at some point in time after the stores issued by the init function pointer have been published.

The only way i see your situation happening would be if the compiler re-ordered instructions past the memory barrier, which it wont do:
https://learn.microsoft.com/en-us/windows/win32/sync/synchronization-and-multiprocessor-issues

**Comment by ioancea** [comment]
@nhorman I know this is counter-intuitive, I had the same push back when I 1st stumbled into ARM64 weakly-ordered memory model. I'll try to explain it a bit.

`MemoryBarrier()` orders memory operations on the CPU that executes it (it's per-thread, not process-wide). It doesn't reach into other CPUs and order their loads.

The reader's fast path in isolation is something like:
```
if (*lock == ONCE_DONE)   // load 1: plain read
 // + later return the shared_data  (load 2: plain read)
```

There is no lock, no barrier, no interlocked op - just two plain loads from two different cache lines. On ARM64, the CPU is free to satisfy `load 2` before `load 1` completes (speculative execution, or stale cache line not yet invalidated). That's the "bug".

The writer's barrier ensures the writer "publishes" in order. But "published" doesn't mean "observed" in the same order. Without a matching barrier on the reader, the reader can observe those two cache lines in any order. I could work on a PoC that simulates this cache-line reordering on Windows ARM64 to demonstrate it empirically if that helps.

~~Ms's (very incomplete) article about this topic https://learn.microsoft.com/en-us/windows/win32/sync/synchronization-and-multiprocessor-issues#fixing-a-race-condition -> their fix uses `InterlockedCompareExchange` on the reader side precisely because the writer's ordering alone isn't sufficient.~~ I just saw you pointed to the same link above :)

**Comment by nhorman** [comment]
Yeah, you're going to have to demonstrate this because I just don't buy it.

Regardless of the scope of memory ordering on a weakly ordered system, its just not applicable here.  Even if the memory barrier only operates on a single cpu core in arm.  The memory barrier on the cou thread that's doing the write guarantees that the write of ONCE_DONE to lock won't complete until all the prior writes issued by the init function pointer complete.  Correspondingly a thread doing a read in this function will never see a value of ONCE_DONE  until such time as the writer thread writes it, at which time (all the init function writes) will have also completed (as they were issued from the CPU that write ONCE_DONE to lock), and done across a write barrier.

I understand that the cpu running the reader thread might reorder loads between lock and the shared data, but the reader won't even issue those shared data load requests until it sees ONCE_DONE from the writer, which as noted is serialized by the memory barrier on the write CPU .

I'm happy to be proven wrong, but you're definitely going to need to demonstrate it for me with some code, because I just don't see it as possible.

**Comment by ioancea** [comment]
Fair enough, I'll try to demonstrate it. Just to be clear, this isn't about right or wrong. It's that ARM64's relaxed memory model invalidates assumptions we've all relied on from x86, where certain ordering guarantees come for free. Whereas on ARM64 they have to be explicitly requested.

Please check the [CRYPTO_THREAD_run_once_fastpath_race.zip](https://github.com/user-attachments/files/29651358/CRYPTO_THREAD_run_once_fastpath_race.zip) which tries to mimic CRYPTO_THREAD_run_once while being used in a tight producer/consumers polling loop, where at each iteration the readers are checking whether the shared data is old (NULL). Before every iteration, the threads and their data (the flag and the global) are reset with a sense barrier, which is also used to fire all the work "at once".

The PoC should print the number of "violations" which, when compiled and run on an Windows ARM64 machine (VM or not), should be > 0. If not, it makes sense to call it several times, or just increase the `ITERATIONS` define (50M, by default). Once the problem is observed, the same executable can be run with the `--fix` parameter (which simply adds the MemoryBarrier after the plain load of `*lock == DONE`), expecting the "violations" to disappear :)

In my case, I run it on an `ARM64 Windows 11, Version 25H2` machine, virtualised from a macOS M1 host, with `4 cores` assigned to it (this is important, the more CPUs the better). Example of my runs:

```
>>>>>> vcvarsarm64.bat
[vcvarsall.bat] Environment initialized for: 'arm64'

>>>>>> build.cmd
CRYPTO_THREAD_run_once_fastpath_race.c

Build OK. Run:
  CRYPTO_THREAD_run_once_fastpath_race.exe                 (reproduce: expect VIOLATIONS > 0)
  CRYPTO_THREAD_run_once_fastpath_race.exe --fix           (reader acquire added: expect 0)

>>>>>> CRYPTO_THREAD_run_once_fastpath_race.exe
CRYPTO_THREAD_run_once reader asymmetry
Reader acquire barrier: OFF (OpenSSL state)
CPUs: 4 | Readers: 3 | Iterations: 50000000

VIOLATIONS: 13
=> REORDERING OBSERVED (NULL glob_register while once==DONE)


>>>>>> CRYPTO_THREAD_run_once_fastpath_race.exe --fix
CRYPTO_THREAD_run_once reader asymmetry
Reader acquire barrier: ON  (fix)
CPUs: 4 | Readers: 3 | Iterations: 50000000

VIOLATIONS: 0
=> No violation
```

**Comment by nhorman** [comment]
@ioancea thank you for that, unfortunately, I don't have a copy of ARM64 Windows, or a suitable machine to emulate its weak memory ordering on ( I don't think we have windows vm's setup on our Mac M1's), So I'm not going to be able to reliably test your reproducer.  I ran it over the weekend on my local windows 11 vm and got no errors, but of course thats a cache coherent intel system, so that says exactly nothing in terms of reliability on weak memory ordering.

**Comment by ioancea** [comment]
@nhorman If you already have a Mac M1, it's a matter of minutes (or hours, since it's Windows 😆 ) to install something like UTM + Windows ARM64 evaluation ISO.
Another alternative is to use a github runner, which now should support windows arm64 for public repositories https://github.com/orgs/community/discussions/155713.

**Comment by nhorman** [comment]
It's a matter of access.  I don't have admin privs on the macs so I need devops assistance to get that installed, and it's a national holiday this weekend for them.

I'll look into the ci runner though

**Comment by nhorman** [comment]
I setup a repo to run your test case, but it seems to be hanging:
https://github.com/nhorman/CRYPTO_THREAD_run_once-test/actions/runs/28814525129/job/85450440735

Its odd, because the same thing happened on my other windows 11 system that I tried this on (which was intel based).  Any idea what might be going on here?

**Comment by ioancea** [comment]
It's probably because of `%*`(see the warning in your run `cl : Command line warning D9024 : unrecognized source file type '%*', object file assumed`), try to remove it.
In a batch file, `%*` expands to "all script arguments", while inside the GitHub Action `run:` block (which is PowerShell or cmd direct) it's passed as literal, which is probably hanging `cl.exe` that might be waiting for input

**Comment by nhorman** [comment]
@ioancea that didn't seem to be it:
https://github.com/nhorman/CRYPTO_THREAD_run_once-test/actions/runs/28826218699/job/85489774729

any other thoughts?

**Comment by ioancea** [comment]
@nhorman I forked your repo and did a few changes (disabled optimization, changed the shell to `cmd`, ...) and got it running - see https://github.com/ioancea/CRYPTO_THREAD_run_once-test/actions/runs/28826792165/job/85491620841

**Comment by ioancea** [comment]
@nhorman I noticed you were also able to run it. Is this enough proof to consider fixing the fast-exit path inside `CRYPTO_THREAD_run_once`?

**Comment by nhorman** [comment]
For me personally?  No, Im still looking through the reproducer, as I don't trust the sync routines yet.  That said, nothing I'm doing should stop you from proposing your change.  Just because I'm not convinced yet shouldn't block you from proposing it, goodness knows I might be wrong on this :)

EDIT: I'm still trying to get my head around the reproducer, and come up with a way to clearly demonstrate this, but heres my thoughts currently:

The reproducer iterates over the following set of steps:

1) run RUN_ONCE
2) reader reads global shared data, checks it for being null, flags an issue if it sees a null pointer
3)  writer waits for all readers to complete, then resets the run_once variable and global shared data
4) repeat

And yes, when this is run, we get a few thousand violations, and for that, your change fixes the issue.

My concern with the test is that step 3 never happens in the real world.  It is, strictly speaking a programming error to reset a run_once variable.  They are meant to be statically initialized, set once, and never change again.  Similarly the shared data should be in a single initial state, get set once, and then effectively be immutable.  i.e. a reader thread should never have any shared data in its cpu cache until the run_once completes, which is guarded by a memory barrier, and after that, any cache reads should be fine, because they never will be mismatched with what is in main memory.  This test violates that assumption by toggling the run_once variable and shared data back to NULL.

I've modified the test to flag the case where such a violation happens on the first iteration, but it never does, leading me to believe that this can only occur at some point after the shared data/run_once variable has been cached from a previous iteration, which never happens with correct usages of run_once.

I'm sorry to be complaining about this so much.  by all rights your proposed change is functionally correct, so I shouldn't be griping here so much.  But I'm concerned about performance.  By making the change you propose, we're going to be adding a MemoryBarrier into every call to CRYPTO_THREAD_run_once in the hot path (i.e. the common case where the run_once routine has already been executed).  We call run_once in lots of places, and do so many times over.
